063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,207 @@
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it later to the reporters.
+
+This proposal will produce an implementation for each type of metrics reporter.
+- `KafkaPrometheusMetricsReporter` usable on brokers (for Kafka metrics) and on Kafka clients (including Connect and Streams)
+- `YammerPrometheusMetricsReporter` usable on brokers (for Yammer metrics)
+
+The Prometheus metrics registry is a singleton and the HTTP server will also be a singleton. This will allow applications to start multiple instances of the reporter (for example in applications containing multiple Kafka clients like Streams, Connect), and still collect all metrics via a single HTTP endpoint per JVM.
+
+The reporter for Kafka metrics will be usable outside of Strimzi by applications using Kafka clients. To do so applications will need to set the `metric.reporters` configuration to `KafkaPrometheusMetricsReporter` and set the reporter configurations accordingly for each Kafka client they instantiate.
+
+![Proposal](./images/063-proposal.png)
+
+Strimzi users currently enable jmx_exporter with the following configuration:
+```
+metricsConfig:
+  type: jmxPrometheusExporter
+  valueFrom:
+    configMapKeyRef:
+      name: kafka-metrics
+      key: kafka-metrics-config.yml
+```
+
+To enable metric reporters, Strimzi users would use a configuration like the following:
+```
+metricsConfig:
+  type: strimziMetricsReporter
+  values:
+    allowList:
+      - "kafka_log.*"
+      - "kafka_network.*"
+```
+
+Strimzi will only allow users to set the `prometheus.metrics.reporter.allowlist` setting via the `allowList` field. The `prometheus.metrics.reporter.listener` will not be customizable in Strimzi and will be set to `http://:9404`. It also helps to avoid conflicts with the Strimzi Kafka Agent currently using port 8080.
+
+### Handling of non-numeric metrics
+
+Out of all the metrics emitted by Kafka brokers and clients, a few of them have non-numeric values. Prometheus only supports numeric values for metrics. When using jmx_exporter it is possible to write rules that move the non-numeric value to a label. For example we do it in [kafka-connect-metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/packaging/examples/metrics/kafka-connect-metrics.yaml#L103-L111) for the `status` metric which has a string value.
+
+I propose to do this automatically in the reporters. For example a metric named: `kafka.connect:type=connector-task-metrics,connector="{connector}",task="{task}"<>status` with the value `running` will be converted into `kafka_connect_connector_task_metrics_status{connector="heartbeats",task="0",status="running"}` and its value will be set to `1.0`.
```

Comment:
I would add just one sentence saying that it's going to expose as many metrics as there are values for the original one. Of course, all metrics will have 0.0 as value. Only one will be 1.0.

Reply:
I'm not sure I understand your comment. What do you mean by `all metrics will have 0.0 as value. Only one will be 1.0.`?

Reply:
Ah ok, it makes more sense. Thanks!